arangod/Replication/InitialSyncer.cpp

```diff
@@ -32,7 +33,10 @@ InitialSyncer::InitialSyncer(ReplicationApplierConfiguration const& configuratio
     : Syncer(configuration), _progress{setter} {}
 
 InitialSyncer::~InitialSyncer() {
-  _batchPingTimer.reset();
+  {
+    MUTEX_LOCKER(guard, _batchPingTimerMutex);
+    _batchPingTimer.reset();
```

Comment:
A destructor must never have to acquire a lock to protect access to member variables. If it does, this is the ultimate sign that there is another race condition, because the object is getting destroyed while some other thread is still accessing it.

Reply:
I think the problem here is that the InitialSyncer schedules some async task which needs to be cancelled as well when deleting the InitialSyncer instance.
There is a race between this task being executed and the destructor code of the InitialSyncer.
IMHO one way to fix this is using such mutex, or, maybe even preferrably, to use shared_from_this and pass the shared_ptr to this into the async task as well.

Reply:
@mpoeter : do you think shared_from_this would be preferrable?

Reply:
The problem is that the mutex in the destructor does not fix the underlying race, it just enforces a sequential order in which the member can be accessed. AFAIK resetting `_batchPingTimer` does not guarantee that the task has finished execution if it has already been scheduled, i.e., if the destructor is executed before the task (because it was faster in acquiring the mutex), the task will access an already destroyed instance which puts us directly into the land of undefined behaviour. So yes, I think shared_from_this is the only viable solution in this case.

Reply:
I tried a mutex-free solution based on `shared_from_this` in the last few minutes, but it doesn't fix the problematic race.
Even when treating the `_batchPingTimer` shared_ptr atomically, there is still the potential for races when the pointer is set to `nullptr` and at the same time the async task rescheduling itself.
IMHO this can only be prevented by using an exclusive lock in both task creation and shutdown. So there is no ambiguity about whether the timer exists or not, and preventing it from rescheduling itself when it shouldn't.

Reply:
I would propose the following approach: the scheduled function captures a `weak_ptr` to the syncer instance. In the function you try to `lock` the `weak_ptr`. If this succeeds we have a safe reference on which we can operate, otherwise the task has been cancelled already.  
This way the function does not prevent the syncer from being reclaimed as long as it is not executed (i.e., avoid circular reference), and we cannot destroy the object while it is being used. What do you think?

Reply:
Mmmh, I don't think using a weak_ptr will fix the problem.
The problem is that the callback may run and schedule itself again after the `_batchPingTimer` has been reset in the destructor. That is what the mutex-based approach is trying to prevent.
It is not only about having an invalid reference to a destroyed `this`.
Consider the following sequence of events:
1) `_batchPingTimer` is set initially, callback task is scheduled
2) destructor starts, resets `_batchPingTimer` to a nullptr
3) callback starts, `this` is still fine. `weak_ptr` can still be acquired. callback schedules itself again
4) destructor finishes

Now while this will prevent accessing an invalid `this`, it will not fix the problem of the callback task being there after the object was destroyed. The scheduler will assert on shutdown in all such cases when there are non-cancelled tasks around on shutdown.

The weak_ptr-based solution does not address this, while the brute-force mutex-based solution does.

Reply:
The mutex-based solution does not prevent that the task and the destructor run concurrently. It only protects access to the `_batchPingTimer`, but the race on `_batchPingTmer` is only a symptom of the race that the instance can be destroyed while the task is still running (and accessing the instance while it is being destroyed).

IMHO the situation you describe is not possible. The destructor is only executed after the shared_ptr's (strong) ref-count reaches zero. But this implies that a call to `weak_ptr.lock` returns null. On the other hand, if `weak_ptr.lock` returns a non-null `shared_ptr`, this means that the task has a safe reference and the object cannot be destroyed as long as the task is still executing. IMHO this should prevent the race on `_batchPingTimer`, and it is not necessary to treat `_bacthPingTimer` atomically, because all memory accesses are synchronized via the updates to the shared_ptr's ref-counts (weak + strong).

I am thinking of something like this:
```cpp
auto weak= this->weak_from_this();
_batchPingTimer = SchedulerFeature::SCHEDULER->queueDelay(
      RequestLane::SERVER_REPLICATION, std::chrono::seconds(secs), [weak](bool cancelled) {
      if (auto self = weak.lock()) {
        // this can only succeed if the (strong) ref-count is non-zero (i.e., the
        // destructor has not yet been executed) and since we now also hold a strong
        // reference, we prevent that the object is destroyed as long as the task has
        // not finished.
        if (!cancelled && self->_batch.id != 0 && !self->isAborted()) {
          self->_batch.extend(self->_state.connection, self->_progress);
          self->startRecurringBatchExtension();
        }
      });
```

Reply:
I tried this solution (with the minor change that I didn't use `weak_from_this` as it is not available in C++14), but it does *not* fix the second problem.
As mentioned before there is also the problem of the task being able to schedule itself while the destructor is running. So `this` still exists and is valid, but the destructor has already set `_batchPingTimer` to a nullptr. So the callback code can still run here and schedule itself again. And then the destructor finishes, leaving the task around. While the task is later safe to execute due to the weak ptr, there is the problem of the Scheduler complaining that there are still tasks around.
But all tasks are expected to be shut down already. This is the problem the mutex-based solution can fix. It can prevent rescheduling the task while the destructor is executing.
In contrast, just using the weak ptr will perfectly fix referring to an invalid Syncer object, but it will not fix having a scheduled task around after the Syncer is destroyed, and this will trigger an assertion failure in the Scheduler on shutdown.

Reply:
@mpoeter: ah yes, I think you are right... :man_facepalming: sometimes I think I am just too slow/dumb.
I implemented the changes you mentioned (actually for the second time now). But now I am convinced that your explanation is right and mine is not, and that actually we don't need the mutex there.
I will start a Jenkins runs with these changes now.

Reply:
I know that feeling - happens to all of us! 🙂 Glad we could sort this out.